concent_api/core/views.py

```diff
@@ -1,223 +1,321 @@
+        last_undelivered_message_status.delivered = True
+        last_undelivered_message_status.save()
+
+    if isinstance(decoded_message_data, MessageForceReportComputedTask):
+        return raw_message_data
+    elif isinstance(decoded_message_data, MessageAckReportComputedTask):
+        message_task_to_compute = load(
+            decoded_message_data.message_task_to_compute,
+            settings.CONCENT_PRIVATE_KEY,
+            client_public_key
+        )
+        if current_time <= message_task_to_compute.deadline + 2 * settings.CONCENT_MESSAGING_TIME:
+            return raw_message_data
         return HttpResponse("", status = 204)
-
-    if decoded_message_data['type'] == "MessageRejectReportComputedTask":
-        message_to_compute          = Message.objects.get(type = 'MessageForceReportComputedTask', task_id = decoded_message_data['message_cannot_commpute_task']['task_id'])
-        raw_message_to_compute      = message_to_compute.data.tobytes()
-        decoded_message_to_compute  = json.loads(raw_message_to_compute.decode('utf-8'))
-        if current_time <= decoded_message_to_compute['message_task_to_compute']['deadline'] + 2 * settings.CONCENT_MESSAGING_TIME:
-            if decoded_message_data['message_cannot_commpute_task']['reason'] == "deadline-exceeded":
-                decoded_message_to_compute['type']      = "MessageAckReportComputedTask"
-                decoded_message_to_compute['timestamp'] = current_time
-                return decoded_message_to_compute
-            return decoded_message_data
+    elif isinstance(decoded_message_data, MessageRejectReportComputedTask):
+        message_cannot_compute_task = load(
+            decoded_message_data.message_cannot_compute_task,
+            settings.CONCENT_PRIVATE_KEY,
+            client_public_key
+        )
+        message_to_compute = Message.objects.get(
+            type = 'MessageForceReportComputedTask',
+            task_id = message_cannot_compute_task.task_id
+        )
+        raw_message_to_compute = message_to_compute.data.tobytes()
+        decoded_message_from_database = load(
+            raw_message_to_compute,
+            settings.CONCENT_PRIVATE_KEY,
+            client_public_key,
+        )
+        message_task_to_compute = load(
+            decoded_message_from_database.message_task_to_compute,
+            settings.CONCENT_PRIVATE_KEY,
+            client_public_key,
+        )
+        if current_time <= message_task_to_compute.deadline + 2 * settings.CONCENT_MESSAGING_TIME:
+            if decoded_message_data.reason == "deadline-exceeded":
+                message_ack_report_computed_task = MessageAckReportComputedTask(
+                    timestamp = current_time,
+                    message_task_to_compute = decoded_message_from_database.message_task_to_compute,
+                )
+                return message_ack_report_computed_task
+            return raw_message_data
         return HttpResponse("", status = 204)
```

Comment:
This is wrong. You cannot just return HTTP 204 if the last message in the queue was discarded. You must process the next message in the queue.